services/appveyor/appveyor.js

```diff
@@ -1,26 +1,23 @@
 'use strict';
 
 const BaseService = require('../base');
+const {
+  checkErrorResponse,
+  asJson,
+} = require('../../lib/error-helper');
 
-/**
- * AppVeyor CI integration.
- */
 module.exports = class AppVeyor extends BaseService {
   async handle({repo, branch}) {
     let apiUrl = 'https://ci.appveyor.com/api/projects/' + repo;
     if (branch != null) {
       apiUrl += '/branch/' + branch;
     }
-    const {buffer, res} = await this._sendAndCacheRequest(apiUrl, {
+    const json = await this._sendAndCacheRequest(apiUrl, {
       headers: { 'Accept': 'application/json' }
-    });
+    }).then(checkErrorResponse.asPromise({ notFoundMessage: 'project not found or access denied' }))
+      .then(asJson);
```

Comment:
Question: If we want to throw an exception here in the `handle()` function because we detect an error condition based on the response content rather than the status code, can you give an example of how to do that? Based on the calling code

```js
     try {
       return await this.handle(namedParams);
     } catch (error) {
       ...
     }
```

it seems like I should be able to `throw new Error('foo');` or `return Promise.reject(new Error('foo'));` in here, but both of those throw `UnhandledPromiseRejectionWarning` if I try that. Obviously once we've got a more final implementation we can document stuff like this.

Reply:
Right! So, the way to signal an exception based on the response content is to `throw new InvalidResponse(message)`, or just `throw new InvalidResponse()` to display **invalid**.

That said, the `UnhandledPromiseRejectionWarning` is a bug (see https://github.com/badges/shields/pull/1590#discussion_r176289473). The behavior I would expect in that situation is that the thrown error is caught and emitted directly. (I _think_ that's possible.)

Reply:
Right yes, so if I add the line `throw new InvalidResponse('foo');` here, that does what we expect. If I add `throw new Error('foo');` here that throws `UnhandledPromiseRejectionWarning`. It seems like that should also be caught in `invokeHandler()`

Reply:
Bit more info: that behaviour is only true with `HANDLE_INTERNAL_ERRORS=false`.

If I set `HANDLE_INTERNAL_ERRORS=true` there, we get a standard 'internal error', but I guess the error it is catching/re-throwing is `UnhandledPromiseRejectionWarning` rather than the `new Error()` we threw.

Reply:
It took me a while to get clarity about what needed to happen here, though I think I've gotten it working well now. Would you like to take a look?